chips/stm32f4xx/src/chip.rs

```diff
@@ -2,39 +2,32 @@
 
 use core::fmt::Write;
 use cortexm4;
-use kernel::Chip;
-
 use kernel::common::deferred_call;
-
-use crate::adc;
-use crate::dma1;
-use crate::exti;
-use crate::fsmc;
-use crate::i2c;
-use crate::nvic;
-use crate::spi;
-use crate::tim2;
-use crate::usart;
+use kernel::Chip;
 
 use crate::deferred_calls::DeferredCallTask;
+use crate::fsmc;
+use crate::interrupt::InterruptService;
 
-pub struct Stm32f4xx {
+pub struct Stm32f4xx<I: InterruptService> {
     mpu: cortexm4::mpu::MPU,
     userspace_kernel_boundary: cortexm4::syscall::SysCall,
     scheduler_timer: cortexm4::systick::SysTick,
+    interrupt_service: I,
 }
 
-impl Stm32f4xx {
-    pub unsafe fn new() -> Stm32f4xx {
+impl<I: InterruptService> Stm32f4xx<I> {
+    pub unsafe fn new(interrupt_service: I) -> Stm32f4xx<I> {
         Stm32f4xx {
             mpu: cortexm4::mpu::MPU::new(),
             userspace_kernel_boundary: cortexm4::syscall::SysCall::new(),
             scheduler_timer: cortexm4::systick::SysTick::new(),
+            interrupt_service,
         }
     }
 }
 
-impl Chip for Stm32f4xx {
+impl<I: InterruptService> Chip for Stm32f4xx<I> {
```

Comment:
Shouldn't this be a separate PR?

Reply:
I was not sure about this. Before this pr, from the service_pending_interrupts() function in stm32f4xx/chip.rs one could only process interrupts that were common to to all stm324xx chips. Because the random number generator stm32f412g specific, in order to use it and receive the associated interrupts, one needs to have access to the nvic id defined in the stm32f412g_nvic.rs file. It would not have made sense for the stm32f412g crate to be imported in the stm32f4xx crate, as that would lead to circular dependencies. By making Chip generic over InterruptService, this allows each variant of the f4 boards to define their own Chip struct where they can process interrupts that are not common to [f4xx](https://github.com/krady21/tock/blob/aafad112e09c41e0f424efca37a03a6bbc2817a0/chips/stm32f412g/src/interrupt.rs#L18-L21). This is not new in tock. Nordic boards do the [same thing](https://github.com/krady21/tock/blob/stm32f4-rng/chips/nrf52840/src/interrupt_service.rs). If needed, I can break this up into two PRs. Let me know.

Reply:
It should probably be a separate PR I think.

Reply:
Yeah, its clearly a desired change but I think that it is kind of obscured by being in what appears to just be a PR for adding a single hardware peripheral to a single chip.

Also, now that #2069 is merged, the `InterruptService` trait is now available in `kernel/src/platform` and has changed slightly. It would be good to use the updated, broadly usable trait rather than an stm specific one.

Reply:
Didn't know about #2069 , so thanks for pointing it out. @hudson-ayers , I see that you also have an ongoing PR (#2187) that implements the interrupt changes for the f3. Do you have another one in the making for the f4 chips and boards? I'm asking in order to avoid overlapping with your work.

Reply:
I do have another PR in the works for the f4 chips and boards, it should be ready within a day or two